src/cpp/session/modules/SessionHTMLPreview.cpp

```diff
@@ -969,34 +980,52 @@ void handlePreviewRequest(const http::Request& request,
    // request for dependent file
    else
    {
-      FilePath filePath = s_pCurrentPreview_->targetDirectory().childPath(path);
+      FilePath filePath;
+      if (!s_currentPageViewerFile.empty())
+         filePath = FilePath(s_currentPageViewerFile).parent().childPath(path);
+      else
+         filePath = s_pCurrentPreview_->targetDirectory().childPath(path);
       addFileSpecificHeaders(filePath, pResponse);
       pResponse->setFile(filePath, request);
    }
 }
 
-SEXP rs_showPageViewer(SEXP fileSEXP, SEXP selfContainedSEXP)
+SEXP rs_showPageViewer(SEXP urlSEXP, SEXP titleSEXP)
 {
    try
    {
-      // get full path to file
-      FilePath filePath = module_context::resolveAliasedPath(r::sexp::safeAsString(fileSEXP));
+      // terminate any running preview
+      if (isPreviewRunning())
+         s_pCurrentPreview_->terminate();
 
-      // get self-contained argument
-      bool selfContained = r::sexp::asLogical(selfContainedSEXP);
+      // get url and title
+      std::string url = r::sexp::safeAsString(urlSEXP);
+      std::string title = r::sexp::safeAsString(titleSEXP);
 
-      // default viewerFilePath to original file path (change it if self-contained)
-      FilePath viewerFilePath = filePath;
+      // paths we will forward via event
+      FilePath filePath, viewerFilePath;
 
-      if (selfContained)
+      // check for url
+      if (boost::algorithm::starts_with(url, "http"))
       {
-         // error if rmarkdown isn't available
-         if (!module_context::isPackageInstalled("rmarkdown"))
+         // verify local url only
+         if (!boost::algorithm::starts_with(url, "http://localhost") &&
```

Comment:
Do we need to check for https here? (It doesn't seem likely, but `mapUrlPorts` considers it possible.)